statsmodels/base/distributed_estimation.py

```diff
@@ -0,0 +1,431 @@
+from statsmodels.base.elastic_net import fit_elasticnet
+from statsmodels.base.model import Results
+from statsmodels.tools.decorators import cache_readonly
+from statsmodels.regression.linear_model import OLS
+#from statsmodels.genmod.generalized_linear_model import GLM
+#from statsmodels.genmod.families import Binomial
+import statsmodels.base.wrapper as wrap
+import numpy as np
+
+"""
+distributed regularized estimation.
+
+Routines for fitting regression models using a distributed
+approach outlined in
+
+Jason D. Lee, Qiang Liu, Yuekai Sun and Jonathan E. Taylor.
+"Communication-Efficient Sparse Regression: A One-Shot Approach."
+arXiv:1503.04337. 2015. http://arxiv.org/abs/1503.04337.
+
+Currently the primary usecase is for out of memory computation,
+this approach currently does not support mutli-core processing
+and processing must be handled sequentially.
+
+
+There are several variables that are taken from the source paper
+for which the interpretation may not be directly clear from the
+code, these are mostly used to help form the estimate of the
+approximate inverse covariance matrix as part of the
+debiasing procedure.
+
+    X_beta
+
+    A weighted design matrix used to perform the node-wise
+    regression procedure.
+
+    gamma_hat
+
+    gamma_hat is produced as part of the node-wise regression
+    procedure used to produce the approximate inverse covariance
+    matrix.  One is produced for each variable using the
+    LASSO.
+
+    tau_hat
+
+    tau_hat is produced using the gamma_hat values for each
+    p to produce weights to reweight the gamma_hat values which
+    are ultimately used to form theta_hat.
+
+    theta_hat
+
+    This is the estimate of the approximate inverse covariance
+    matrix.  This is used to debiase the coefficient average
+    along with the average gradient.
+"""
+
+def _gen_grad(mod, params, n_obs, alpha, L1_wt, score_kwds):
+    """generates the log-likelihood gradient for the debiasing
+
+    Parameters
+    ----------
+    mod : statsmodels model class
+        The model for the current machine.
+    params : array-like
+        The estimated coefficients for the current machine.
+    n_obs : scalar
+        machine specific sample size.
+    alpha : scalar or array-like
+        The penalty weight.  If a scalar, the same penalty weight
+        applies to all variables in the model.  If a vector, it
+        must have the same length as `params`, and contains a
+        penalty weight for each coefficient.
+    L1_wt : scalar
+        The fraction of the penalty given to the L1 penalty term.
+        Must be between 0 and 1 (inclusive).  If 0, the fit is
+        a ridge fit, if 1 it is a lasso fit.
+    score_kwds : dict-like or None
+        Keyword arguments for the score function.
+
+    Returns
+    -------
+    An array-like object of the same dimension as params
+
+    Notes
+    -----
+    In general:
+
+    nabla l_k(params)
```

Comment:
I would use "gradient" instead of "nabla" and clarify what the "k" denotes.  Also, the log-likelihood whould not include the 1/n_obs factor.